modules/rgbd/src/volume.cpp

```diff
@@ -5,16 +5,75 @@
 #include "precomp.hpp"
 #include <opencv2/rgbd/volume.hpp>
 
-#include "tsdf.hpp"
 #include "hash_tsdf.hpp"
+#include "tsdf.hpp"
 
 namespace cv
 {
 namespace kinfu
 {
-cv::Ptr<Volume> makeVolume(VolumeType _volumeType, float _voxelSize, cv::Matx44f _pose,
-                           float _raycastStepFactor, float _truncDist, int _maxWeight,
-                           float _truncateThreshold, Vec3i _resolution)
+
+Ptr<VolumeParams> VolumeParams::defaultParams(VolumeType _volumeType)
+{
+    VolumeParams params;
+    params.type              = _volumeType;
+    params.maxWeight         = 64;
+    params.raycastStepFactor = 0.25f;
+    params.unitResolution = 0;         // unitResolution not used for TSDF
+    float volumeSize         = 3.0f;
+    params.pose = Affine3f().translate(Vec3f(-volumeSize / 2.f, -volumeSize / 2.f, 0.5f));
+    switch (params.type)
+    {
+        case VolumeType::TSDF:
+            params.resolution          = Vec3i::all(512);
+            params.voxelSize           = volumeSize / 512.f;
+            params.depthTruncThreshold = 0.f;  // depthTruncThreshold not required for TSDF
+            break;
+        case VolumeType::HASHTSDF:
+            params.unitResolution      = 16;
+            params.voxelSize           = volumeSize / 512.f;
+            params.depthTruncThreshold = rgbd::Odometry::DEFAULT_MAX_DEPTH();
+        default:
+            //! TODO: Should throw some exception or error
+            break;
+    }
+    params.tsdfTruncDist = 7 * params.voxelSize;  //! About 0.04f in meters
+
+    return makePtr<VolumeParams>(params);
+}
+
+Ptr<VolumeParams> VolumeParams::coarseParams(VolumeType _volumeType)
+{
+    Ptr<VolumeParams> params = defaultParams(_volumeType);
+
+    params->raycastStepFactor = 0.75f;
+    float volumeSize          = 3.0f;
+    switch (params->type)
+    {
+        case VolumeType::TSDF:
+            params->resolution = Vec3i::all(128);
+            params->voxelSize  = volumeSize / 128.f;
+            break;
+        case VolumeType::HASHTSDF: params->voxelSize = volumeSize / 128.f; break;
+        default: break;
+    }
+    params->tsdfTruncDist = 2 * params->voxelSize;  //! About 0.04f in meters
+    return params;
+}
+
+Ptr<Volume> makeVolume(const VolumeParams& _volumeParams)
+{
+    switch (_volumeParams.type)
+    {
+        case VolumeType::TSDF: return kinfu::makeTSDFVolume(_volumeParams); break;
+        case VolumeType::HASHTSDF: return makePtr<kinfu::HashTSDFVolumeCPU>(_volumeParams); break;
```

Comment:
Why here `makePtr` is used to create a `Volume` while the function below uses `makeHashTSDFVolume`? Shouldn't it be changed to the same way?

Reply:
Added another function for this reason. Done.